src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -115,13 +115,45 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
             var otherAsHashSet = collection as HashSet<T>;
             if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
             {
-                _buckets = (int[])otherAsHashSet._buckets.Clone();
-                _slots = (Slot[])otherAsHashSet._slots.Clone();
+                int count = otherAsHashSet._count;
```

Comment:
The body of this if is really long.  Perhaps it should be extracted to `InitializeFromOtherHashSet` (I'm not sold on the name).

Then it just looks like

```C#
if (isCompatbleHashSet)
{
    InitializeFromOtherHashSet(otherHashSet);
}
else
{
    // Maybe this should also be a method.
    InitializeFromEnumerable(collection);
}

if (needsToTrim)
{
    TrimExcess();
}
```